File: ruby/part2/chap7.rb
class SynchronizeXXXData
  attr_reader :is_update_ready, :is_synch_completed, :is_cache_enabled

  def initialize is_update_ready, is_synch_completed, is_cache_enabled
    @is_update_ready = is_update_ready
    @is_synch_completed = is_synch_completed
    @is_cache_enabled = is_cache_enabled
  end

  def update_database is_force_update
    data_update = ["Ruby", "PHP", "iOS", "Android"]
    if is_update_ready
      if is_force_update
        if !is_synch_completed
          update_backup_db
        else
          update_main_db data_update
          update_backup_db
        end
      else
        update_cache
      end
    end
  end

  private
  def update_backup_db
    p "update backup db"
  end

  def update_main_db data_update
    if !data_update.empty?
      p "parse data to object model"
      p "check validation"
      p "update main database"
    end
  end

  def update_cache
    if is_cache_enabled
      p "disable cache"
      p "update cache"
      p "output message update cache success"
    else
      p "output message errors"
    end
  end
end

test = SynchronizeXXXData.new true, false, true
test.update_database true

File: php/part2/Chap7.php
<?php

class SynchronizeXXXData
{
    public $isUpdateReady;
    public $isSynchCompleted;
    public $isCacheEnabled;

    public function __construct($isUpdateReady = false, $isSynchCompleted = false, $isCacheEnabled = false)
    {
        $this->$isUpdateReady = $isUpdateReady;
        $this->$isSynchCompleted = $isSynchCompleted;
        $this->$isCacheEnabled = $isCacheEnabled;
    }

    public function updateDatabase($isForcusUpdate)
    {
        $dataUpdate = [
            'PHP',
            'Ruby',
            'IOS',
            'Java',
        ];

        if ($isUpdateReady) {
            if ($isForcusUpdate) {
                if (! $isSynchCompleted) {
                    $this->updateBackupDB();
                } else {
                    $this->updateMainDB($dataUpdate);
                    $this->updateBackupDB();
                }
            } else {
                $this->updateCache();
            }
        }
    }

    private function updateBackupDB()
    {
        echo "update backup db";
    }

    private updateMainDB($dataUpdate = [])
    {
        if (! empty($dataUpdate)) {
            echo "parse data to object model";
            echo "check validation";
            echo "update main database";
        }
    }

    private function  updateCache()
    {
        if ($isCacheEnabled) {
            echo "disable cache";
            echo "update cache";
            echo "output message update cache success";
        } else {
            echo "output message errors";
        }
    }
}
